pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -271,19 +272,39 @@ void ConsumerImpl::messageReceived(const ClientConnectionPtr& cnx, const proto::
 
     unsigned int numOfMessageReceived = 1;
     if (metadata.has_num_messages_in_batch()) {
-        Lock lock(mutex_);
```

Comment:
A. Yes I need the mutex since batchAcknowledgementTracker_ is not thread safe
B. As discussed - I don't see a deadlock